Documentation/ceph-recover-pod-on-LostNode.md

```diff
@@ -0,0 +1,138 @@
+Normal   SuccessfulAttachVolume  3m6s               attachdetach-controller                AttachVolume.Attach succeeded for volume "pvc-ccb1108e-4d3d-4478-9c16-56a1e3001a7a"
+Warning  FailedMount             94s (x8 over 19m)  kubelet, ip-10-0-159-129.ec2.internal  Unable to attach or mount volumes: unmounted volumes=[mypvc], unattached volumes=[mypvc default-token-s2fdk]: timed out waiting for the condition
+Normal   AddedInterface          78s                multus                                 Add eth0 [10.129.2.21/23]
+Normal   Pulling                 78s                kubelet, ip-10-0-159-129.ec2.internal  Pulling image "docker.io/library/nginx:latest"
+Normal   Pulled                  74s                kubelet, ip-10-0-159-129.ec2.internal  Successfully pulled image "docker.io/library/nginx:latest"
+Normal   Created                 74s                kubelet, ip-10-0-159-129.ec2.internal  Created container web-server
+Normal   Started                 74s                kubelet, ip-10-0-159-129.ec2.internal  Started container web-server
+
+```
+
+# Node lost and still has the device in it
+
+### In this case, you will see pod in `ContainerCreating` state
+
+```console
+kubectl -n rook-ceph get pod
+```
+>```
+> csi-rbd-demo-depl-69cd495f9b-cqbcd           0/1     ContainerCreating   0          24s
+>```
+
+And, when you will describe the pod `Warning` like below
+
+```console
+$ kubectl -n rook-ceph describe pod csi-rbd-demo-depl-69cd495f9b-cqbcd
+
+Warning  FailedMount             115s (x3 over 6m28s)  kubelet, ip-10-0-159-129.ec2.internal  Unable to attach or mount volumes: unmounted volumes=[mypvc], unattached volumes=[mypvc default-token-s2fdk]: timed out waiting for the condition
+Warning  FailedMount             114s (x7 over 7m25s)  kubelet, ip-10-0-159-129.ec2.internal  MountVolume.MountDevice failed for volume "pvc-ccb1108e-4d3d-4478-9c16-56a1e3001a7a" : rpc error: code = Internal desc = rbd image replicapool/csi-vol-1a1bbb34-7461-11eb-a32b-0a580a81020c is still being used
+```
+
+**Solution:**
+
+The solution is to remove the watcher, following the commands below
+
+```console
+$ kubectl -n rook-ceph exec -it rook-ceph-tools-6b4889fdfd-2qrmt -- sh
+```
+```console
+rbd ls --pool=replicapool
+```
+>```
+> csi-vol-1a1bbb34-7461-11eb-a32b-0a580a81020c
+>```
+
+```console
+rbd status csi-vol-1a1bbb34-7461-11eb-a32b-0a580a81020c --pool=replicapool
+```
+>```
+> Watchers:
+>	watcher=10.130.2.1:0/2076971174 client.14206 cookie=18446462598732840961
+>```
+
+```console
+$ ceph osd blacklist add 10.130.2.1:0/207697117 # to know which watcher to block see above output
+blacklisting 10.130.2.1:0/2076971174 until 2021-02-21T18:12:53.115328+0000 (3600 sec)
+```
+
+After running the above command within a few minutes the pod will be running. **But don't forget to remove the above blacklist watcher**
```

Comment:
If the blacklist is removed, this could cause corruption if that node comes back at a later time and the volume reconnects. So we should be very clear that the blacklisting can only be removed if the node is guaranteed to be gone.

Reply:
Two scenarios here.

* when Node is completely dead

once the backlist is removed, bring back the original node and see we are able to get the pod running on the lost node.

* When only control-plane is lost but data-plane is still active

once the backlist is removed, start the kubelet on the node and see pod is running and we are able to access the data.

Reply:
Those two scenarios are about getting the original node running again? Doesn't the admin need to go through those steps to recover a node before they try to force delete the pod and get it running somewhere else? If the blacklist is removed, it is removing the check that prevents corruption so I thought the admin could only do that if they're guaranteeing the node is permanently dead, at least at the data-plane.

Reply:
> Those two scenarios are about getting the original node running again? Doesn't the admin need to go through those steps to recover a node before they try to force delete the pod and get it running somewhere else?

Blacklisting helps to recover the application pod, If the pod is still active on the node where control-plane is lost we can fence it off so no read/write happens.

 If the blacklist is removed, it is removing the check that prevents corruption so I thought the admin could only do that if they're guaranteeing the node is permanently dead, at least at the data-plane.

Yes, we need to test it out and confirm if the node comes back how it's going to behave.